File: apps/android/app/src/main/java/com/voiid/app/main/clips/ClipTextOverlay.kt
package com.voiid.app.main.clips

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Typeface
import android.os.Build
import android.text.Layout
import android.text.StaticLayout
import android.text.TextPaint
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shadow
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.core.content.res.ResourcesCompat
import com.voiid.app.R
import com.voiid.app.ui.theme.VoiidFont
import java.util.UUID
import kotlin.math.ceil

/**
 * Text placed on a clip in the editor, and the one drawing of it the export burns in — so the
 * words land in the uploaded video where they sat on screen. Port of iOS ClipTextOverlay.swift.
 *
 * ONE LAYOUT, TWO RENDERERS. Everything is measured against the VIDEO's width, never the
 * screen's: the size is a fraction of the frame width and the position a fraction of the
 * frame. The editor lays the text out over the video's on-screen rect with those fractions;
 * the export draws the same fractions into the full-resolution frame, so a 1080-wide export
 * and a 390dp-wide phone agree.
 */
data class ClipTextOverlay(
    val id: String = UUID.randomUUID().toString(),
    val text: String,
    val color: Int = 0,
    val pill: Boolean = false,
    /** Centre of the text, 0…1 of the video frame, y down. */
    val x: Float = 0.5f,
    val y: Float = 0.4f,
) {
    val fill: Int get() = PALETTE[color.coerceIn(0, PALETTE.lastIndex)]
    /** The letters' colour: the palette colour, or on a pill the colour that reads on it. */
    val ink: Int get() = if (pill) (if (color == 0) android.graphics.Color.BLACK else android.graphics.Color.WHITE) else fill

    companion object {
        /** Font size as a fraction of the frame width: 28 on a 390-wide frame. */
        const val FONT_FRACTION = 28f / 390f
        /** Longest line before wrapping, as a fraction of the frame width. */
        const val WRAP_FRACTION = 300f / 390f
        const val MAX_LENGTH = 120

        val PALETTE = intArrayOf(
            android.graphics.Color.WHITE,
            android.graphics.Color.BLACK,
            android.graphics.Color.rgb(19, 130, 140),
            android.graphics.Color.rgb(255, 214, 51),
            android.graphics.Color.rgb(255, 97, 140),
            android.graphics.Color.rgb(89, 153, 255),
        )
    }
}

/** Draws the overlays into a transparent frame-sized bitmap for the exporter. */
object ClipTextRenderer {
    private fun typeface(context: Context): Typeface {
        val base = runCatching { ResourcesCompat.getFont(context, R.font.nunito_variable) }.getOrNull()
            ?: Typeface.DEFAULT
        return if (Build.VERSION.SDK_INT >= 28) Typeface.create(base, 700, false)
        else Typeface.create(base, Typeface.BOLD)
    }

    /** Null when there is nothing to draw. */
    fun overlay(context: Context, texts: List<ClipTextOverlay>, width: Int, height: Int): Bitmap? {
        val visible = texts.filter { it.text.isNotBlank() }
        if (visible.isEmpty() || width <= 0 || height <= 0) return null
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        val face = typeface(context)
        visible.forEach { draw(canvas, it, width, height, face) }
        return bitmap
    }

    private fun draw(canvas: Canvas, item: ClipTextOverlay, width: Int, height: Int, face: Typeface) {
        val size = width * ClipTextOverlay.FONT_FRACTION
        val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            typeface = face
            textSize = size
            color = item.ink
            if (!item.pill) setShadowLayer(size * 0.14f, 0f, size * 0.035f, android.graphics.Color.argb(115, 0, 0, 0))
        }
        val wrap = (width * ClipTextOverlay.WRAP_FRACTION).toInt()
        // Measure the widest line so the pill hugs the text instead of spanning the wrap width.
        val probe = StaticLayout.Builder.obtain(item.text, 0, item.text.length, paint, wrap)
            .setAlignment(Layout.Alignment.ALIGN_CENTER).build()
        val lineWidth = (0 until probe.lineCount).maxOf { probe.getLineWidth(it) }
        val layoutWidth = ceil(lineWidth).toInt().coerceIn(1, wrap)
        val layout = StaticLayout.Builder.obtain(item.text, 0, item.text.length, paint, layoutWidth)
            .setAlignment(Layout.Alignment.ALIGN_CENTER).build()

        val cx = item.x * width
        val cy = item.y * height
        val left = cx - layout.width / 2f
        val top = cy - layout.height / 2f

        if (item.pill) {
            val padX = size * 0.5f
            val padY = size * 0.21f
            val r = size * 0.36f
            val bg = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = item.fill }
            canvas.drawRoundRect(RectF(left - padX, top - padY, left + layout.width + padX,
                top + layout.height + padY), r, r, bg)
        }
        canvas.save()
        canvas.translate(left, top)
        layout.draw(canvas)
        canvas.restore()
    }
}

/** The same text in Compose, sized for a video that is [frameWidth] wide on screen. */
@Composable
fun ClipTextLabel(item: ClipTextOverlay, frameWidth: Dp, modifier: Modifier = Modifier) {
    val density = LocalDensity.current
    val sizeDp = frameWidth * ClipTextOverlay.FONT_FRACTION
    // Pinned to the frame, not the user's font scale: the export has no font scale.
    val sizeSp = with(density) { (sizeDp.toPx() / (density.density * density.fontScale)) }
    val style = VoiidFont.rounded(sizeSp, FontWeight.Bold).let {
        if (item.pill) it else it.copy(shadow = Shadow(Color(0x73000000),
            Offset(0f, with(density) { (sizeDp * 0.035f).toPx() }), with(density) { (sizeDp * 0.14f).toPx() }))
    }
    Text(
        item.text,
        style = style,
        color = Color(item.ink),
        textAlign = TextAlign.Center,
        modifier = modifier
            .widthIn(max = frameWidth * ClipTextOverlay.WRAP_FRACTION + if (item.pill) sizeDp else Dp(0f))
            .then(
                if (item.pill) Modifier
                    .background(Color(item.fill), RoundedCornerShape(sizeDp * 0.36f))
                    .padding(horizontal = sizeDp * 0.5f, vertical = sizeDp * 0.21f)
                else Modifier
            ),
    )
}

File: apps/ios/Voiid/Voiid/Main/Clips/ClipTextOverlay.swift
//
//  ClipTextOverlay.swift
//  Voiid
//
//  Text placed on a clip in the editor, and the one drawing of it that both the editor and
//  the export use — so the words land in the uploaded video exactly where they sat on screen.
//
//  ── ONE LAYOUT, TWO RENDERERS ───────────────────────────────────────────────────
//  Everything is measured against the VIDEO's width, never the screen's: the size is a
//  fraction of the frame width and the position a fraction of the frame. The editor lays the
//  text out over the video's on-screen rect with those fractions; the export draws the same
//  fractions into the full-resolution frame. A 1080-wide export and a 390-point phone
//  therefore agree, which a point size would not.
//

import SwiftUI
import UIKit
import CoreImage

nonisolated struct ClipTextOverlay: Identifiable, Equatable, Sendable {
    enum Style: Equatable { case plain, pill }

    let id = UUID()
    var text: String
    var color: Int = 0
    var style: Style = .plain
    /// Centre of the text, 0…1 of the video frame, y down.
    var position = CGPoint(x: 0.5, y: 0.4)

    /// Font size as a fraction of the frame width: 28 pt on a 390 pt-wide frame.
    static let fontFraction: CGFloat = 28.0 / 390.0
    /// Longest line before wrapping, as a fraction of the frame width.
    static let wrapFraction: CGFloat = 300.0 / 390.0

    static let palette: [UIColor] = [
        .white, .black, UIColor(red: 0.075, green: 0.51, blue: 0.55, alpha: 1),
        UIColor(red: 1, green: 0.84, blue: 0.2, alpha: 1), UIColor(red: 1, green: 0.38, blue: 0.55, alpha: 1),
        UIColor(red: 0.35, green: 0.6, blue: 1, alpha: 1),
    ]

    var fill: UIColor { Self.palette[min(color, Self.palette.count - 1)] }
    /// The letters' colour: the palette colour, or on a pill the colour that reads on it.
    var ink: UIColor { style == .pill ? (color == 0 ? .black : .white) : fill }

    static func font(forFrameWidth width: CGFloat) -> UIFont {
        let size = width * fontFraction
        let base = UIFont.systemFont(ofSize: size, weight: .bold)
        guard let rounded = base.fontDescriptor.withDesign(.rounded) else { return base }
        return UIFont(descriptor: rounded, size: size)
    }
}

// MARK: - Export drawing

nonisolated enum ClipTextRenderer {
    /// All the text as one transparent image the size of the frame, in Core Image space
    /// (origin bottom-left) ready to composite over each video frame. Nil when there is none.
    static func overlay(_ texts: [ClipTextOverlay], frame size: CGSize) -> CIImage? {
        let visible = texts.filter { !$0.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
        guard !visible.isEmpty, size.width > 0, size.height > 0 else { return nil }

        let format = UIGraphicsImageRendererFormat()
        format.scale = 1
        format.opaque = false
        let image = UIGraphicsImageRenderer(size: size, format: format).image { _ in
            for item in visible { draw(item, in: size) }
        }
        guard let cg = image.cgImage else { return nil }
        // UIKit drew top-down; CIImage(cgImage:) keeps the picture upright, so no flip.
        return CIImage(cgImage: cg)
    }

    private static func draw(_ item: ClipTextOverlay, in size: CGSize) {
        let font = ClipTextOverlay.font(forFrameWidth: size.width)
        let paragraph = NSMutableParagraphStyle()
        paragraph.alignment = .center
        var attrs: [NSAttributedString.Key: Any] = [
            .font: font, .foregroundColor: item.ink, .paragraphStyle: paragraph,
        ]
        if item.style == .plain {
            let shadow = NSShadow()
            shadow.shadowColor = UIColor.black.withAlphaComponent(0.45)
            shadow.shadowBlurRadius = font.pointSize * 0.14
            shadow.shadowOffset = CGSize(width: 0, height: font.pointSize * 0.035)
            attrs[.shadow] = shadow
        }
        let text = NSAttributedString(string: item.text, attributes: attrs)
        let wrap = size.width * ClipTextOverlay.wrapFraction
        let bounds = text.boundingRect(with: CGSize(width: wrap, height: .greatestFiniteMagnitude),
                                       options: [.usesLineFragmentOrigin, .usesFontLeading], context: nil)
            .integral
        let centre = CGPoint(x: item.position.x * size.width, y: item.position.y * size.height)
        let textRect = CGRect(x: centre.x - bounds.width / 2, y: centre.y - bounds.height / 2,
                              width: bounds.width, height: bounds.height)

        if item.style == .pill {
            let padX = font.pointSize * 0.5, padY = font.pointSize * 0.21
            let pill = textRect.insetBy(dx: -padX, dy: -padY)
            item.fill.setFill()
            UIBezierPath(roundedRect: pill, cornerRadius: font.pointSize * 0.36).fill()
        }
        text.draw(with: textRect, options: [.usesLineFragmentOrigin, .usesFontLeading], context: nil)
    }
}

// MARK: - Editor drawing

/// The same text in SwiftUI, sized for a video that is `frameWidth` points wide on screen.
struct ClipTextLabel: View {
    let item: ClipTextOverlay
    let frameWidth: CGFloat

    var body: some View {
        let font = ClipTextOverlay.font(forFrameWidth: frameWidth)
        Text(item.text)
            .font(Font(font))
            .multilineTextAlignment(.center)
            .foregroundColor(Color(item.ink))
            .shadow(color: .black.opacity(item.style == .plain ? 0.45 : 0),
                    radius: font.pointSize * 0.14, y: font.pointSize * 0.035)
            .frame(maxWidth: frameWidth * ClipTextOverlay.wrapFraction)
            .fixedSize(horizontal: false, vertical: true)
            .padding(.horizontal, item.style == .pill ? font.pointSize * 0.5 : 0)
            .padding(.vertical, item.style == .pill ? font.pointSize * 0.21 : 0)
            .background(
                RoundedRectangle(cornerRadius: font.pointSize * 0.36, style: .continuous)
                    .fill(item.style == .pill ? Color(item.fill) : .clear)
            )
    }
}
